### PR TITLE
HPCC-17192 Refactor PTree to have atom and non-atom versions

### PR DESCRIPTION
**Pull request by jakesmith**
Previously all property trees used atom tables for attribute
names, value and node names.

Now there are two implementations, one which uses atoms as before
and the other does not.

The atom version continues to be used by Dali for memory
conversatin reasons.  All other uses now by default use the
non-atom version.

The non-atom version is significantly quicker, in particular in
multi-threaded use cases, because it does not have to enter/leave
the mutext involved in the global atom tables.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [x] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:

+ Regression suite
+ datest
+ switch all to/from either variety of ptree
+ valgrind memcheck

### CODE REVIEW

**Comment by jakesmith** [comment]
@ghalliday - please review

File: dali/base/dacsds.ipp [hunk 1]
```diff
@@ -333,8 +335,8 @@ public:
     virtual void appendLocal(size32_t size, const void *data, bool binary);
     virtual void addingNewElement(IPropertyTree &child, int pos);
     virtual void removingElement(IPropertyTree *tree, unsigned pos);
-    virtual void setAttr(const char *attr, const char *val);
-    virtual bool removeAttr(const char *attr);
+    virtual void setAttribute(const char *attr, const char *val);
```
Comment (richardkchapman) [inline dali/base/dacsds.ipp]: Worth adding override here?

Comment (jakesmith) [inline dali/base/dacsds.ipp]: yep, I like override, it's saved me in the past. TBH I'd forgotten to use it recently.

File: dali/base/dacsds.cpp [hunk 2]
```diff
@@ -52,19 +52,19 @@ static CriticalSection SDScrit;
 
 ////////////////////
 
-class MonitoredChildMap : public ChildMap
+class MonitoredChildMap : public ChildMapAtom
```
Comment (ghalliday) [inline dali/base/dacsds.cpp]: naming: It matches the existing convention (NC), but it is hard to read.  In english it is normal to specialise a class name by prefixing the class name with the specialism.  i.e. AtomChildMap.  Otherwise it reads as a specialised type of Atom.

File: system/jlib/jptree.cpp [hunk 3]
```diff
@@ -3176,6 +2818,408 @@ bool PTree::checkPattern(const char *&xxpath) const
...
+    if (oname)
+        kT->releaseKey(oname);
+}
+
+const char *CAtomPTree::queryName() const
+{
+    return name?name->get():nullptr;
+}
+
+AttrValue *CAtomPTree::newAttrArray(unsigned n)
```
Comment (ghalliday) [inline system/jlib/jptree.cpp]: Do we significantly gain from this mechanism?  If so should it also be being used by the non-commoned code?

File: system/jlib/jptree.cpp [hunk 4]
```diff
@@ -3176,6 +2818,408 @@ bool PTree::checkPattern(const char *&xxpath) const
...
+            {
+                del = a;
+                break;
+            }
+        }
+    }
+    if (!del)
+        return false;
+    numAttrs--;
+    CriticalBlock block(hashcrit);
```
Comment (ghalliday) [inline system/jlib/jptree.cpp]: This is a shame... removing an attribute needs to lock and clone all the attributes.

File: dali/base/dacsds.cpp (hunk 2 above)
Comment (jakesmith) [inline dali/base/dacsds.cpp]: ok will change.

**Review by ghalliday — COMMENTED** [review]
I'm not sure that using a different representation for the non-atom tables is worthwhile, and increasing the number of base classes makes life complicated.

I would be more inclined to go the other way and combine the attribute map into the PTree base class - partly because it may then allow the class to be correctly packed and avoid the current misaligned accesses.

@richardkchapman thoughts?

**Comment by richardkchapman** [comment]
I agree about the confusing class structure...

File: system/jlib/jptree.cpp [hunk 5]
```diff
@@ -3176,6 +2818,408 @@ bool PTree::checkPattern(const char *&xxpath) const
     return ret;
 }
 
+//////////////////////
+
+unsigned ChildMapC::getHashFromElement(const void *e) const
+{
+    LocalPTree &elem = (LocalPTree &) (*(IPropertyTree *)e);
+    const char *name = elem.queryName();
+    return hashc((const unsigned char *)name, (size32_t)strlen(name), 0);
```
Comment (ghalliday) [inline system/jlib/jptree.cpp]: It looks like I lost a comment.

Have you considered the efficiency impact of repeating this hash?  Particularly when hash tables are expanded the cost can be quite high.

Simpler would be to always create an "atom", but only common it up in dali.  It would avoid creating AtomChildMap etc..

File: system/jlib/jptree.ipp [hunk 6]
```diff
@@ -404,14 +338,240 @@ private:
...
+    }
+    virtual bool matchesFindParam(const void *e, const void *fp, unsigned fphash) const override
+    {
+        return (0 == stricmp(((IPropertyTree *)e)->queryName(), (const char *)fp));
+    }
+};
+
+
+
+class jlib_decl CAtomPTree : public PTree
```
Comment (ghalliday) [inline system/jlib/jptree.ipp]: I seem to have lost a comment here as well.

I'm not keen on the duplication of the structure.  E.g., having two completely different ways of allocating attribute arrays.  It would be simpler to always create AttrStrs and store those in the lists - the difference being that one was commoned up, the other left as it is?

**Comment by ghalliday** [comment]
@jakesmith I think I should have assigned this back to you

**Comment by jakesmith** [comment]
@ghalliday - please review

File: system/jlib/jptree.ipp [hunk 7]
```diff
@@ -403,15 +370,160 @@ private:
...
-    HashKeyElement *name;
     ChildMap *children;
     IPTArrayValue *value;
-    //The packing (#pragma pack) is overridden because very large numbers of these objects are created, and the
-    //following two members currently cause 8 bytes to be wasted.  Refactoring the contents of AttrMap into this
-    //class would allow the fields to pack cleanly.
-    AttrMap attributes;     // this has 2 "extra" bytes - which could pack into the space following the count
-    byte flags;             // this could also pack into the space following the count.
+    byte flags;
+    HashKeyElement *name = nullptr;
```
Comment (ghalliday) [inline system/jlib/jptree.ipp]: New jira: Should open a new jira to review removing the packing from this structure - reordering the entries should enable the size to be the same, but access to be much more efficient.

File: system/jlib/jptree.cpp [hunk 8]
```diff
@@ -3176,6 +2813,261 @@ bool PTree::checkPattern(const char *&xxpath) const
...
+    {
+        if (streq(v->value->get(), val))
+            return;
+        AttrStrC::destroy((AttrStrC *)v->value);
+        v->value = AttrStrC::create(val);
+    }
+    else
+    {
+        attrs = (AttrValue *)realloc(attrs, (numAttrs+1)*sizeof(AttrValue));
+        if (isnocase())
```
Comment (ghalliday) [inline system/jlib/jptree.cpp]: future: I suspect there is some scope for moving functions from AttrStrNC/C into AttrStr - e.g., create(key, isnocase()) and hash/destroy.

File: system/jlib/jptree.cpp [hunk 9]
```diff
@@ -3176,6 +2813,261 @@ bool PTree::checkPattern(const char *&xxpath) const
...
+    return true;
+}
+
+void LocalPTree::setAttribute(const char *key, const char *val)
+{
+    if (!key)
+        return;
+    if (!validateXMLTag(key+1))
+        throw MakeIPTException(-1, "Invalid xml attribute: %s", key);
+    if (!val)
```
Comment (ghalliday) [inline system/jlib/jptree.cpp]: future: How common is this?  If common it  might be worth having a cached AttrStrC value.

File: system/jlib/jptree.cpp [hunk 10]
```diff
@@ -3176,6 +2813,261 @@ bool PTree::checkPattern(const char *&xxpath) const
...
+    if (v)
+    {
+        if (streq(v->value->get(), val))
+            return;
+        AttrStrC::destroy((AttrStrC *)v->value);
+        v->value = AttrStrC::create(val);
+    }
+    else
+    {
+        attrs = (AttrValue *)realloc(attrs, (numAttrs+1)*sizeof(AttrValue));
```
Comment (ghalliday) [inline system/jlib/jptree.cpp]: future: Consider increasing this in powers of two, and using an efficient isPowerOfTwo test to check if it needs to reallocate.

File: system/jlib/jptree.ipp [hunk 11]
```diff
@@ -242,59 +241,6 @@ struct AttrValue
 // Ideally the classes would be restructured to avoid this, but it would probably require AttrMap to move into PTree
```
Comment (ghalliday) [inline system/jlib/jptree.ipp]: You can now delete
AM_NOCASE_FLAG, AM_NOCASE_MASK
just above this line in the header

**Review by ghalliday — APPROVED** [review]
@jakesmith some suggestions for future improvments, but I think they can all go in future PRs.

@richardkchapman please double check and merge if happy.

File: system/jlib/jptree.ipp [hunk 12]
```diff
@@ -332,69 +277,91 @@ public:
     IPTArrayValue *detachValue() { IPTArrayValue *v = value; value = NULL; return v; }
     void setValue(IPTArrayValue *_value, bool binary) { if (value) delete value; value = _value; if (binary) IptFlagSet(flags, ipt_binary); }
     bool checkPattern(const char *&xxpath) const;
-    void clear();
+    IPropertyTree *detach()
+    {
+        IPropertyTree *tree = create(queryName(), value, children, true);
+        PTree *_tree = QUERYINTERFACE(tree, PTree); assertex(_tree); _tree->setParent(this);
+
+        unsigned short _numAttrs = _tree->numAttrs;
```
Comment (jakesmith) [inline system/jlib/jptree.ipp]: changing as part of this PR

File: system/jlib/jptree.ipp (hunk 11 above)
Comment (jakesmith) [inline system/jlib/jptree.ipp]: Removing as part of this PR

File: system/jlib/jptree.ipp (hunk 7 above)
Comment (jakesmith) [inline system/jlib/jptree.ipp]: Have tested size difference.
As it stands now with packing, PTree size is 55 bytes
Without packing and members moved suitably, size is 56 bytes.

So worth doing and since this PR combined the classes, I'll make this change as part of this PR too.

**Comment by jakesmith** [comment]
@ghalliday - changed pushed, please check.

Let me know when you're done and I'll squash.

File: system/jlib/jptree.ipp [hunk 13]
```diff
@@ -332,115 +268,280 @@ public:
...
-    HashKeyElement *name;
-    ChildMap *children;
-    IPTArrayValue *value;
-    //The packing (#pragma pack) is overridden because very large numbers of these objects are created, and the
-    //following two members currently cause 8 bytes to be wasted.  Refactoring the contents of AttrMap into this
-    //class would allow the fields to pack cleanly.
-    AttrMap attributes;     // this has 2 "extra" bytes - which could pack into the space following the count
-    byte flags;             // this could also pack into the space following the count.
+    /* NB: the order of the members here is important to reduce the size of the objects, because very large numbers of these are created.
+     * The base CInterfaceOf contains it's VMT + a unsigned (atom).
```
Comment (ghalliday) [inline system/jlib/jptree.ipp]: It is a unsigned link count, rather than an atom, same in line below.

**Comment by jakesmith** [comment]
@ghalliday - fixed comments and squashed.

**Comment by ghalliday** [comment]
@richardkchapman ready for second review and merging

File: dali/base/dasds.cpp [hunk 14]
```diff
@@ -1624,7 +1624,7 @@ class CXMLFileExternal : public CExternalFile, implements IExternalHandler
     virtual void resetAsExternal(IPropertyTree &_tree)
     {
         PTree &tree = *QUERYINTERFACE(&_tree, PTree);
-        tree.clear();
+        Owned<IPropertyTree> toDelete = tree.detach();
```
Comment (richardkchapman) [inline dali/base/dasds.cpp]: Is this equivalent to ::Release(tree.detach());

And if so, would it not have been clearer to code it that way?

**Comment by richardkchapman** [comment]
Seems to be a spurious change to cpp-driver submodule

**Comment by richardkchapman** [comment]
@jakesmith Back to you

File: dali/base/dasds.cpp (hunk 14 above)
Comment (jakesmith) [inline dali/base/dasds.cpp]: It is equivalent.. I'm used to this form.
I don't mind changing it though.

**Comment by jakesmith** [comment]
@richardkchapman - addressed comments.
